### DISCUSSION TITLE
Reseting the layout via watcher: accessing OS window properties

### DISCUSSION THREAD

**Discussion by anjos** [kovidgoyal/kitty#6466] [General]
I wanted to automatically switch between layouts depending on the window width.  A narrower window should use the "vertical" layout, while a wider window should use the "tall" configuration (preset like this: `tall;bias=50;full_size=1;mirrored=false`).

I thought I should do this with a watcher:

```python
import typing

from kitty.boss import Boss
from kitty.window import Window


def on_resize(
    boss: Boss, window: Window, data: typing.Mapping[str, typing.Any]
) -> None:
    """Kitty callback when the window is resized

    If the window width is larger than a certain preset amount, then we use the
    "tall" layout.  Otherwise, if it is too think, we use the vertical layout.
    """
    # Here data will contain old_geometry and new_geometry
    ng = data["new_geometry"]
    if (ng.right - ng.left) >= 3000:
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "tall")
        )
    else:
        boss.call_remote_control(window, ("goto-layout", "--match", "all", "vertical"))
```

I configured the watcher globally, on `kitty.conf`.   This seems to work well, when the window is narrower (`vertical` layout is set), but not when I trigger the `tall` layout.  At that point, creating a new window starts making kitty unstable: it seems to be going from one layout to the other endlessly, very quickly.  The only way to stop this is to exit any newly created window.

I think that the problem is that if I create a new window on the "tall" layout as configured, it will immediately trigger the change in geometry for the same window again, which will make it go back to the "vertical" layout, but then the width of the window will be wider again and the trigger will happen once more, and that endlessly.

I guess I want to measure the change in geometry on the OS Window instead of the kitty window - but how to achieve this in a nice way?

Thanks for any tips!

**Comment by anjos** [comment]
Another bit of info you may know about: Ideally, I wanted to measure the number of "columns" in the layout change.  For example: if the resized window has more than 160 columns, then I'd know it can fit two windows side-by-side.  This would also take into account the current font size.  Thanks!

**Comment by kovidgoyal** [comment]
Use viewport_for_window(window.os_window_id)

On Fri, Jul 14, 2023 at 02:46:18PM -0700, André Anjos wrote:
> I wanted to automatically switch between layouts depending on the window width.  A narrower window should use the "vertical" layout, while a wider window should use the "tall" configuration (preset like this: `tall;bias=50;full_size=1;mirrored=false`).
> 
> I thought I should do this with a watcher:
> 
> ```python
> import typing
> 
> from kitty.boss import Boss
> from kitty.window import Window
> 
> 
> def on_resize(
>     boss: Boss, window: Window, data: typing.Mapping[str, typing.Any]
> ) -> None:
>     """Kitty callback when the window is resized
> 
>     If the window width is larger than a certain preset amount, then we use the
>     "tall" layout.  Otherwise, if it is too think, we use the vertical layout.
>     """
>     # Here data will contain old_geometry and new_geometry
>     ng = data["new_geometry"]
>     if (ng.right - ng.left) >= 3000:
>         boss.call_remote_control(
>             window, ("goto-layout", "--match", "all", "tall")
>         )
>     else:
>         boss.call_remote_control(window, ("goto-layout", "--match", "all", "vertical"))
> ```
> 
> I configured the watcher globally, on `kitty.conf`.   This seems to work well, when the window is narrower (`vertical` layout is set), but not when I trigger the `tall` layout.  At that point, creating a new window starts making kitty unstable: it seems to be going from one layout to the other endlessly, very quickly.  The only way to stop this is to exit any newly created window.
> 
> I think that the problem is that if I create a new window on the "tall" layout as configured, it will immediately trigger the change in geometry for the same window again, which will make it go back to the "vertical" layout, but then the width of the window will be wider again and the trigger will happen once more, and that endlessly.
> 
> I guess I want to measure the change in geometry on the OS Window instead of the kitty window - but how to achieve this in a nice way?
> 
> Thanks for any tips!
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6466
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by anjos** [comment]
Got that and the final implementation looks like this:

```python
import typing

from kitty.boss import Boss
from kitty.cli import create_default_opts
from kitty.fast_data_types import viewport_for_window
from kitty.window import Window


def on_resize(
    boss: Boss, window: Window, data: typing.Mapping[str, typing.Any]
) -> None:
    """Kitty callback when the window is resized

    We switch the layout based on the total number of cells (symbols) we have
    horizontally, going from the vertical (default), tall (with one left-size
    full-sized window), and then tall and finally horizontal.
    """

    # We first retrieve the default width for the window, which in my case
    # represents a single terminal session:
    default_width = create_default_opts().initial_window_width[0]

    # We add 1 cell to account for borders
    threshold = default_width + 1

    # We first estimate the width of the window.  It is not precise as the
    # window width include a couple of pixels to the left and right.  We
    # use the
    _, _, window_width, _, cell_width, _ = viewport_for_window(window.os_window_id)
    cells = window_width / cell_width
    if cells < (2 * threshold):
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "vertical")
        )
    elif cells < (3 * threshold):
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "tall")
        )
    else:
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "horizontal")
        )
```

Last question: I read the configuration options to determine the threshold for switching between layouts.  Is there a better way to read the configuration option `initial_window_width` than `kitty.cli.create_default_opts().initial_window_width`?  I'm asking because I don't need to (re-)load the whole configuration, just to access the initial value set.  May be something from the `boss` allows me to access those options w/o reloading?

**Comment by kovidgoyal** [comment]
On Sat, Jul 15, 2023 at 03:24:07AM -0700, André Anjos wrote:
> Last question: I read the configuration options to determine the threshold for switching between layouts.  Is there a better way to read the configuration option `initial_window_width` than `kitty.cli.create_default_opts().initial_window_width`?  I'm asking because I don't need to (re-)load the whole configuration, just to access the initial value set.  May be something from the `boss` allows me to access those options w/o reloading?

get_options().whatever

**Comment by anjos** [comment]
That worked, thanks.  Final script for posterity here:

```python
import typing

from kitty.boss import Boss
from kitty.cli import create_default_opts
from kitty.fast_data_types import get_options, viewport_for_window
from kitty.window import Window


def on_resize(
    boss: Boss, window: Window, data: typing.Mapping[str, typing.Any]
) -> None:
    """Kitty callback when the window is resized

    We switch the layout based on the total number of cells (symbols) we have
    horizontally, going from the vertical (default), tall (with one left-size
    full-sized window), and then tall (with 2 left-size full-sized window),
    when we have the maximum space available (e.g. full screen on a monitor).

    Configuration is:

    enabled_layouts vertical,tall:full_size=1;bias=50;mirrored=false,tall:full_size=2;bias=66;mirrored=false,horizontal
    """

    # We first retrieve the default width for the window, which in my case
    # represents a single terminal session:
    default_width = get_options('initial_window_width')[0]

    # We add 1 cell to account for borders
    threshold = default_width + 1

    # We first estimate the width of the window.  It is not precise as the
    # window width include a couple of pixels to the left and right.  We
    # use the
    _, _, window_width, _, cell_width, _ = viewport_for_window(window.os_window_id)
    cells = window_width / cell_width
    if cells < (2 * threshold):
        boss.call_remote_control(window, ("goto-layout", "--match", "all", "vertical"))
    elif cells < (3 * threshold):
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "tall:full_size=1")
        )
    elif cells < (4 * threshold):
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "tall:full_size=2")
        )
    else:
        boss.call_remote_control(
            window, ("goto-layout", "--match", "all", "horizontal")
        )
```

**Reply by anjos** [reply]
Of course, I misread what @kovidgoyal wrote.  The line with `get_options()` above should actually read `default_width = get_options().initial_window_width[0]`.
